### PR TITLE
feat: safe-fs and path validator

### PR DESCRIPTION
**Pull request by zeyu2001**
This adds a new package, `@opengovsg/starter-kitty-fs`, and a new validator to `@opengovsg/starter-kitty-validators`.

## safe-fs

Creates a safe-by-default version of the Node.js `fs` module.

```typescript
safeFs: (basePath?: string) => typeof fs
```

This is a safe version of the Node.js `fs` module, guarded against path traversal attacks and absolute path usage.

All interfaces are exactly the same as the original `fs` module, but all file paths are resolved relative to the provided `basePath` and are guaranteed to fall within the `basePath` when the fs operations are executed.

The use of file descriptors in place of a file path is not affected.

## createPathSchema

Create a schema that validates user-supplied pathnames for filesystem operations.

```typescript
createPathSchema: (options?: PathValidatorOptions) => ZodSchema<string>
```

This is added to the `@opengovsg/starter-kitty-validators` package.

### CODE REVIEW

File: packages/validators/src/path/options.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+   * @example `'/var/www'`
+   */
+  basePath?: string
+}
+
+export const optionsSchema = z.object({
+  basePath: z
+    .string()
+    .optional()
+    .default(() => process.cwd())
```
Comment (spaceraccoon) [inline packages/validators/src/path/options.ts]: hmm tbh I'd prefer this not be optional so users declaratively decide what's a safe path; and in some cases `process.cwd` is usually unsafe e.g. containing `.env` or source code

File: packages/validators/src/path/utils.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  if (untrustedPath.indexOf('\0') !== -1) {
+    return false
+  }
+  // check for backslashes
+  if (untrustedPath.indexOf('\\') !== -1) {
+    return false
+  }
+  // resolve the path relative to the Node process's current working directory
+  // since that's what fs operations will be relative to
+  const normalizedPath = path.resolve(untrustedPath) // normalizedPath is now an absolute path
```
Comment (spaceraccoon) [inline packages/validators/src/path/utils.ts]: Seems like `path.resolve` is run twice - another time in the `.transform()` step in `schema.ts`. Is this intentional?

File: packages/validators/src/path/utils.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  // check for dot segments, even if they don't normalize to anything
+  if (normalizedPath.includes('..')) {
+    return false
+  }
+
+  // check if the normalized path is within the provided 'safe' base path
+  if (normalizedPath.indexOf(basePath) !== 0) {
+    return false
+  }
```
Comment (spaceraccoon) [inline packages/validators/src/path/utils.ts]: This works but still a bit uncomfy as we should use Node's native APIs more to ensure the output is as close to how Node would actually interpret it later e.g. if dev uses `fs.readFileSync` or something. Probably some combination of `[path.relative](https://nodejs.org/api/path.html#pathrelativefrom-to)`? 

```
> path.relative('/safe/foo','/evil/bar')
'../../evil/bar'
> path.relative('/safe/foo','/safe/foo/bar')
'bar'
> path.relative('/safe/foo','../foo/bar')
'../../Users/user1/Documents/foo/bar'
> path.relative('/safe/foo','foo/bar')
'../../Users/user1/Documents/test/foo/bar'
```

File: packages/validators/src/__tests__/path.test.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import path from 'node:path'
+
+import { describe, expect, it } from 'vitest'
+import { ZodError } from 'zod'
+
+import { OptionsError } from '@/common/errors'
+import { createPathSchema } from '@/index'
+
+describe('Path validator with default options', () => {
```
Comment (spaceraccoon) [inline packages/validators/src/__tests__/path.test.ts]: Very naise

File: packages/safe-fs/src/params.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
+{
+  "access": [0],
```
Comment (spaceraccoon) [inline packages/safe-fs/src/params.json]: Me being super annoying but does it make sense to have this as a `.ts` export instead and an object literal? Then you can add comments on each fnName so the argument locations are visible to the reader and avoid some of the jank of importing from a JSON.

File: packages/safe-fs/src/getter.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      typeof target[p] === 'function' &&
+      // required for func to get the correct type
+      p !== 'promises' &&
+      p !== 'constants' &&
+      p !== 'Stats' &&
+      p !== 'StatsFs' &&
+      p !== 'Dirent' &&
+      p !== 'Dir' &&
+      p !== 'ReadStream' &&
+      p !== 'WriteStream'
```
Comment (spaceraccoon) [inline packages/safe-fs/src/getter.ts]: Note that this is a workaround for TypeScript to properly parse this as a function and not <any> function, but maybe there's a better TypeScript-friendly way to do this

**Review by spaceraccoon — COMMENTED** [review]
Very cool use of Proxy and Reflect sir - minor comments and maybe we check tomorrow whether there's a better way to handle the getter type

File: packages/safe-fs/src/getter.ts (hunk 6 above)
Comment (zeyu2001) [reply packages/safe-fs/src/getter.ts]: Used `Exclude<T, U>` to do this, but still somewhat manual:

```typescript
type FsFunctionNames = Exclude<
  keyof typeof fs,
  | 'promises'
  | 'constants'
  | 'Stats'
  | 'StatsFs'
  | 'Dirent'
  | 'Dir'
  | 'ReadStream'
  | 'WriteStream'
>
```

Comment (zeyu2001) [reply packages/safe-fs/src/getter.ts]: Ok TIL about `Extract`: `type FsFunction = Extract<(typeof fs)[keyof typeof fs], CallableFunction>`

**Pull request merged**